File: backend/api/schemas/config_test.py
"""Config-test harness schemas (W43 F2, ADR-0040 §Decision 3).

The on-platform试跑 harness: preview a DRAFT (unsaved) per-KB retrieval/citation
config by running the full `/query` pipeline N times and aggregating presentation
counters (citation / figure raw+dedup) + latency with a variance band — so the KB
owner sees the config's effect before persisting it (and can A/B vs the saved one).
"""

from typing import Literal

from pydantic import BaseModel, Field


class DraftRetrievalConfig(BaseModel):
    """The 12 W43 per-KB knobs under test (draft, unsaved). Mirrors the `KbConfig`
    Optional knobs + `EffectiveConfig` / `PerQueryOverrides` field shape. Every field
    is Optional — `None` = fall through to the KB's saved config / global default in
    the resolver (so a draft only overrides the knobs it sets)."""

    enable_parent_doc_retrieval: bool | None = None
    parent_doc_section_depth_offset: int | None = None
    parent_doc_top_k: int | None = None
    parent_doc_max_tokens_per_parent: int | None = None
    enable_citation_post_hoc_expansion: bool | None = None
    citation_expansion_max_aux: int | None = None
    citation_expansion_window: int | None = None
    citation_expansion_section_path_prefix_depth: int | None = None
    enable_citation_neighbour_images: bool | None = None
    citation_neighbour_max_aux_images: int | None = None
    citation_neighbour_section_path_prefix_depth: int | None = None
    max_images_per_answer: int | None = None


class ConfigTestRequest(BaseModel):
    """POST /kb/{kb_id}/config-test input."""

    query: str = Field(..., min_length=1, max_length=2000)
    # N runs for the variance band — 證偽實驗④ saw ~20% figure swing run-to-run, so a
    # single run can't decide a config. Capped at 5 (each run = a full synth, ~10-40s).
    runs: int = Field(default=3, ge=1, le=5)
    draft_config: DraftRetrievalConfig = Field(default_factory=DraftRetrievalConfig)
    mode: Literal["hybrid", "vector", "fulltext"] = "hybrid"
    # Default CRAG off for a clean per-config signal (CRAG re-synth adds its own variance).
    enable_crag: bool = False
    # F2.4 — also run N with the SAVED config (all-draft knobs cleared) for an A/B.
    compare_to_saved: bool = False


class CitationBreakdown(BaseModel):
    """Per-citation section + image count (F2.3 — 每 citation section+圖數)."""

    chunk_id: str
    section_path: list[str]
    image_count: int


class RunMetrics(BaseModel):
    """Presentation counters + latency for one run."""

    run: int
    citation_count: int
    figure_count_raw: int  # total embedded_images across citations (cross-citation dups kept)
    figure_count_dedup: int  # unique by checksum_sha256 (matches the frontend display dedup)
    latency_ms: int
    answer_chars: int
    refused: bool


class MetricBand(BaseModel):
    """Aggregate of one metric across runs. `band` = max - min = the variance band
    (per 證偽發現④ — if the band swamps a draft-vs-saved difference, the signal is
    noise; F2.6 trust gate checks exactly this)."""

    min: float
    max: float
    mean: float
    band: float


class ConfigRunSummary(BaseModel):
    """N-run aggregate for one config (draft or saved)."""

    runs: list[RunMetrics]
    citation_count: MetricBand
    figure_count_raw: MetricBand
    figure_count_dedup: MetricBand
    latency_ms: MetricBand
    per_citation: list[CitationBreakdown]  # from the last run (representative)


class ConfigTestResult(BaseModel):
    """POST /kb/{kb_id}/config-test output."""

    kb_id: str
    query: str
    runs: int
    # The resolved EffectiveConfig actually applied (draft > saved KbConfig > global)
    # — lets the caller confirm what was tested.
    resolved_config: dict[str, int | bool | None]
    draft: ConfigRunSummary
    saved: ConfigRunSummary | None = None  # populated when compare_to_saved=True

File: frontend/lib/api/config-test.ts
/**
 * Config-test harness API (W43 F2/F3.3, per ADR-0040 §Decision 3).
 *
 * The on-platform 試跑 harness: preview a DRAFT (unsaved) per-KB retrieval/citation
 * config by running the full `/query` pipeline N times and aggregating presentation
 * counters (citation / figure raw+dedup) + latency with a variance band. Mirrors
 * `backend/api/schemas/config_test.py` field-for-field.
 */

import { ApiClient } from '../api-client';

const client = new ApiClient();

/**
 * The 12 W43 per-KB knobs under test (draft, unsaved). Every field Optional —
 * `null`/absent = fall through to the KB's saved config / global default in the
 * resolver (so a draft only overrides the knobs it sets). Mirrors `KbConfig`'s
 * Optional knobs + backend `DraftRetrievalConfig`.
 */
export interface DraftRetrievalConfig {
  enable_parent_doc_retrieval?: boolean | null;
  parent_doc_section_depth_offset?: number | null;
  parent_doc_top_k?: number | null;
  parent_doc_max_tokens_per_parent?: number | null;
  enable_citation_post_hoc_expansion?: boolean | null;
  citation_expansion_max_aux?: number | null;
  citation_expansion_window?: number | null;
  citation_expansion_section_path_prefix_depth?: number | null;
  enable_citation_neighbour_images?: boolean | null;
  citation_neighbour_max_aux_images?: number | null;
  citation_neighbour_section_path_prefix_depth?: number | null;
  max_images_per_answer?: number | null;
}

export interface ConfigTestRequest {
  query: string;
  /** N runs for the variance band (1–5, default 3). Each run is a full synth. */
  runs?: number;
  draft_config?: DraftRetrievalConfig;
  mode?: 'hybrid' | 'vector' | 'fulltext';
  /** Default false — CRAG re-synth adds its own variance to a per-config signal. */
  enable_crag?: boolean;
  /** Also run N with the SAVED config for an A/B (F2.4). */
  compare_to_saved?: boolean;
  /**
   * W48 (ADR-0040 dual-axis) — also compute the reference-free RAGAs faithfulness
   * quality axis (judge LLM). Default true server-side; set false for a fast
   * presentation-only run. Degrades to null when no judge credential is configured.
   */
  eval_faithfulness?: boolean;
}

/** Per-citation section + image count (one row of the breakdown). */
export interface CitationBreakdown {
  chunk_id: string;
  section_path: string[];
  image_count: number;
}

/** Presentation counters + latency for one run. */
export interface RunMetrics {
  run: number;
  citation_count: number;
  figure_count_raw: number;
  figure_count_dedup: number;
  latency_ms: number;
  answer_chars: number;
  refused: boolean;
}

/** Aggregate of one metric across runs. `band` = max − min = the variance band. */
export interface MetricBand {
  min: number;
  max: number;
  mean: number;
  band: number;
}

/** N-run aggregate for one config (draft or saved). */
export interface ConfigRunSummary {
  runs: RunMetrics[];
  citation_count: MetricBand;
  figure_count_raw: MetricBand;
  figure_count_dedup: MetricBand;
  latency_ms: MetricBand;
  per_citation: CitationBreakdown[];
  /**
   * W48 (ADR-0040 dual-axis) → W49 (決策 7) — reference-free RAGAs faithfulness for
   * this config, judged PER RUN and aggregated into a MetricBand (band = max − min)
   * so the quality axis exposes its run-to-run noise. N=1 → band=0 (the card then
   * shows a single-shot warning). null = quality axis skipped (eval_faithfulness=false)
   * or no judge / every run's judge errored.
   */
  faithfulness: MetricBand | null;
}

export interface ConfigTestResult {
  kb_id: string;
  query: string;
  runs: number;
  /** The resolved EffectiveConfig actually applied (draft > saved KbConfig > global). */
  resolved_config: Record<string, number | boolean | null>;
  draft: ConfigRunSummary;
  saved: ConfigRunSummary | null;
}

export const configTestApi = {
  // POST /kb/{kb_id}/config-test — runs the full pipeline N times against a draft
  // config (and optionally the saved config) without persisting anything.
  run: (kbId: string, body: ConfigTestRequest): Promise<ConfigTestResult> =>
    client.post<ConfigTestResult>(`/kb/${kbId}/config-test`, body),
};
